### DISCUSSION TITLE
Using cloud-init

### DISCUSSION THREAD

**Discussion by Nordal** [bpg/terraform-provider-proxmox#952] [Q&A]
Hi all

I would like to use a real cloud-init user_data file, and not use the limited proxmox build-in options.
It seems to be doable, but the documentation and examples maybe could be better. 

I have made a snippets enabled datastore on proxmox called "snippets" and made the following terraform code:

in the initialization resource, i have added this line:

  user_data_file_id = proxmox_virtual_environment_file.ci-user_data.id

and made the following file resource:

resource "proxmox_virtual_environment_file" "ci-user_data" {
  content_type = "snippets"
  datastore_id = "snippets"
  node_name    = "proxnode01"

  source_file {
    path = "ci-user_data.yaml"
  }
}

I have placed the ci-user_data.yaml in the root of the terraform folder, but it's not working.

Could anyone come with an example on how to do this?
I have earlier been working with KVM directly, and it was so easy creating a cd-rom containing the user_data file and everything was working. Maybe I am blinded by the "old" method and not seeing the forest because of all the trees.

Looking forward to hear from you.
BR Kasper

**Comment by otopetrik** [comment]
Hi @Nordal,

[this](https://gist.github.com/otopetrik/f6e6ef79636592cab957407da1ac5f7b) worked for me, it might be a useful starting point.

Using cloud-init with the default `bios="seabios"` generally works.

Using cloud-init with `bios="ovmf"` seems to require:
```
initialization {
    interface = "scsi1" # or scsi2, scsi3,...
...
}
```
But there are some, who have [no issues](https://github.com/bpg/terraform-provider-proxmox/issues/586#issuecomment-1826288326) with the `ide2` default value. It might depend on the image used - maybe a Linux image with different cloud-init version or configuration.

Make sure to read about the [guest agent](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#qemu-guest-agent), Proxmox behavior with agent enabled and not installed is a bit surprising.

**Comment by Nordal** [comment]
Thank you, I will have a look.
I am deploying by cloning a template in proxmox and not by an ISO file. I assume i can skip most of the declarations in the bottom.

**Reply by otopetrik** [reply]
That `content_type = "iso"` is required for Proxmox to accept disk image upload (it does not like `.raw` or `.img`), but it is not an CD/DVD ISO file, it is just a normal cloud-init-compatible disk image (which usually resizes to the requested disk size on first boot).

There should not be anything preventing clones working with cloud-init, but I have not used clones in a long time.
When using a clone, the terraform file does not completely describe the VM, it does not have VM configuration inherited from the template. That makes it easy to miss something (though `terraform plan` and `terraform show` help).

Not sure if template VM should have the cloud-init drive already created, probably not, but that is just a guess.
